lib/core/src/irods_default_paths.cpp

```diff
@@ -18,31 +21,83 @@ namespace irods {
         try {
             boost::filesystem::path path{dl_info.dli_fname};
             path = boost::filesystem::canonical(path);
-            path.remove_filename().remove_filename().remove_filename(); // Removes filename and the two directories (usr and lib) between libirods_common.so and base of irods install
+            path.remove_filename();
             return path;
         } catch(const boost::filesystem::filesystem_error& e) {
             THROW(-1, e.what());
         }
     }
 
+    boost::filesystem::path
+    get_irods_root_directory() {
+        try {
+            boost::filesystem::path path = get_irods_lib_directory();
+            boost::filesystem::path install_libdir{IRODS_DEFAULT_PATH_LIBDIR};
+            install_libdir = install_libdir.lexically_normal();
+            long install_path_len = std::distance(install_libdir.begin(), install_libdir.end());
+            if (install_libdir.is_absolute()) {
+                install_path_len--;
+            }
+            for (long i = 0; i < install_path_len; i++) {
+                path.remove_filename(); // Removes directories (usually usr and lib) between libirods_common.so and base of irods install
+            }
+            return path;
+        } catch(const boost::filesystem::filesystem_error& e) {
+            THROW(-1, e.what());
+        }
+    }
+
+    // we can tolerate a little re-prefixing if we ignore common base directories.
+    // For example, if libdir is "usr/lib" and bindir is "usr/bin", we can ignore
+    // "usr" when traversing from one to the other.
+    static inline boost::filesystem::path
+    get_irods_directory_impl(boost::filesystem::path install_path) {
+        // if input path is absolute, do not traverse
+        if (install_path.is_absolute()) {
+            return install_path;
+        }
+        boost::filesystem::path path = get_irods_lib_directory();
+        boost::filesystem::path install_libdir{IRODS_DEFAULT_PATH_LIBDIR};
+        install_libdir = install_libdir.lexically_normal();
+        // if install_libdir is absolute, we can't be smart about it
+        if (install_libdir.is_absolute()) {
+            for (auto ptr = ++install_libdir.begin(); ptr != install_libdir.end(); ptr++) {
+                path.remove_filename();
+            }
+            return path / install_path;
+        }
+        boost::filesystem::path install_path_rel = install_path.lexically_relative(install_libdir);
+        // if there is no common base, we can't be smart about it
+        if (install_path_rel.empty()) {
+            for (auto ptr = install_libdir.begin(); ptr != install_libdir.end(); ptr++) {
+                path.remove_filename();
+            }
+            return path / install_path;
+        }
+
+        path = path / install_path_rel;
```

Comment:
This can be replaced by `path /= install_path_rel;`. The `/=` operator may lead to better machine code (probably not the most important thing to worry about as this code is probably not used in any hot path).